src/lib.rs

```diff
@@ -103,10 +103,33 @@
 //! [`Slab::with_capacity`]: struct.Slab.html#with_capacity
 
 use std::iter::{FromIterator, IntoIterator};
+use std::num::NonZeroUsize;
 use std::ops;
 use std::vec;
 use std::{fmt, mem};
 
+/// Encode the specified index.
+///
+/// We encode the index as a `NonZeroUsize` (which also means that we need
+/// to avoid zero) because it gives entries which can be optimized for zero
+/// sized types like `()`.
+#[inline(always)]
+fn index_encode(index: usize) -> NonZeroUsize {
+    // TODO: Consider restructuring internal index plumbing to avoid this panic.
+    // Safety: This check is necessary to to make sure we're using
+    // `new_unchecked` safely.
+    assert!(index != std::usize::MAX, "index `{}` out of bounds", index);
+    unsafe { NonZeroUsize::new_unchecked(index + 1) }
```

Comment:
I think the `unsafe` can be removed while avoiding redundant checks:
```suggestion
    NonZeroUsize::new(index + 1).expect("index is out of bounds")
```

Reply:
The suggested code is not equivalent. `assert!(index != std::usize::MAX)` is about the overflow of `index + 1`.

Reply:
`usize` is unsigned, so it would overflow to zero, which would be then caught by `NonZeroUsize::new`.

Reply:
Oh, I'm sorry, that's fine in release mode.
However, in debug mode, this will panic for another reason so I think we need to keep the original assertion as a debug assertion.

Reply:
Or use `index.wrapping_add(1)`